polars/polars-arrow/src/kernels/ewm/average.rs

```diff
@@ -1,270 +1,177 @@
-use std::fmt::Debug;
-use std::ops::AddAssign;
-
+use arrow::array::PrimitiveArray;
+use arrow::types::NativeType;
 use num::Float;
 
-use crate::trusted_len::{PushUnchecked, TrustedLen};
+use crate::trusted_len::PushUnchecked;
+
 // See:
 // https://github.com/pola-rs/polars/issues/2148
 // https://stackoverflow.com/a/51392341/6717054
 
-// this is the adjusted variant in pandas
-pub fn ewma_no_nulls<T, I>(vals: I, alpha: T) -> Vec<T>
+pub fn ewm_mean<T>(
+    xs: &PrimitiveArray<T>,
+    alpha: T,
+    adjust: bool,
+    min_periods: usize,
+) -> PrimitiveArray<T>
 where
-    T: Float + AddAssign,
-    I: IntoIterator<Item = T>,
-    I::IntoIter: TrustedLen,
+    T: Float + NativeType,
 {
-    let mut iter = vals.into_iter();
-    let len = iter.size_hint().1.unwrap();
-    if len == 0 {
-        return vec![];
-    }
-    let mut weight = T::one();
-    let mut out = Vec::with_capacity(len);
+    let n = xs.len();
+    let mut results = Vec::with_capacity(n);
 
-    let first = iter.next().unwrap();
-    out.push(first);
-    let mut ewma_old = first;
+    let mut denom = T::zero();
     let one_sub_alpha = T::one() - alpha;
+    let mut non_null_cnt = 0usize;
 
-    let mut i = T::from(out.len()).unwrap();
-    for val in iter {
-        weight += one_sub_alpha.powf(i);
-        ewma_old = ewma_old * (one_sub_alpha) + val;
-        // Safety:
-        // we allocated vals.len()
-        unsafe { out.push_unchecked(ewma_old / weight) };
-        i += T::one();
-    }
-
-    out
-}
-
-pub fn ewma<T, I>(vals: I, alpha: T) -> (usize, Vec<T>)
-where
-    T: Float + AddAssign,
-    I: IntoIterator<Item = Option<T>>,
-    I::IntoIter: TrustedLen,
-{
-    let mut iter = vals.into_iter();
-    let len = iter.size_hint().1.unwrap();
-    if len == 0 {
-        return (0, vec![]);
-    }
-    let mut weight = T::one();
-    let mut out = Vec::with_capacity(len);
+    let mut opt_ewma = None;
 
-    let leading_null_count = set_first_none_null(&mut iter, &mut out);
-    let mut ewma_old = out[out.len() - 1];
-    let one_sub_alpha = T::one() - alpha;
-
-    let mut i = T::one();
-    let mut prev = out[out.len() - 1];
-    for opt_val in iter {
-        prev = match opt_val {
-            Some(val) => {
-                weight += one_sub_alpha.powf(i);
-                ewma_old = ewma_old * (one_sub_alpha) + val;
-                i += T::one();
-                ewma_old / weight
+    for opt_x in xs.iter() {
+        if opt_ewma.is_none() && opt_x.is_some() {
```

Comment:
There are more branches than strictly needed. I believe you can replace line 30 to 49 with:

```rust
        if let Some(&x) = opt_x {
            opt_ewma = opt_ewma.map(|prev_ewma| {
                let value = if adjust {
                    let numer = prev * denom * one_sub_alpha + x;
                    denom = T::one() + one_sub_alpha * denom;
                    numer / denom
                } else {
                    x * alpha + prev * one_sub_alpha
                };
                value
            })
        }
```

Reply:
Neat! I had to modify your suggestion slightly, but the resulting code is much cleaner than my original implementation